### PR TITLE
`Programming exercises`: Add git-diff generation between template and solution repositories

### PR DESCRIPTION
**Pull request by Hialus**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
In order to create code hints for behavioral test cases, we will need the difference between the template and solution repositories in addition to a test coverage. This is required, so we do not generate code hints for code that is already given in the template.

### Description

This PR introduces the generation of such a difference utilizing the git-diff tool.
It uses JGit to generate the git-diff between the template and solution repositories and then converts it to a single ProgrammingExerciseGitDiffReport with one ProgrammingExerciseGitDiffEntry per change.
In order to save space in the database these diff entries do *not* contain the actual code changes. They only contain the file paths, start lines of the code blocks and the length of the code blocks (in lines).
ProgrammingExerciseFullGitDiffReportDTO and ProgrammingExerciseFullGitDiffEntryDTO are derived from these and *do* contain the actual code blocks. They are created from the light diff-report in ProgrammingExerciseGitDiffService.

The light git-diff is created every time the template or solution repositories get updated. The full git-diff is created only when needed by the client. (currently only when the client requests it with a button in the detail view) This is also the reason they are merely DTOs.

**UML Diagram:**

grafik

**Example:**
The git-diff output
```
--- src/test.java
+++ src/test.java
@@ -1,1 +1,3 @@
- // TODO: Add class
+ public class Test {
+     // TODO: Implement class
+ }
```
would result in the following (light) DiffEntry
```
previousFilePath: "src/test.java",
filePath: "src/test.java",
previousStartLine: 1,
startLine: 1,
previousLineCount: 1,
lineCount: 3
```
and the following FullDiffEntry
```
previousFilePath: "src/test.java",
filePath: "src/test.java",
previousLine: 1,
line: 1,
previousCode: "// TODO: Add class",
code: """
public class Test {
    // TODO: Implement class
}
"""
```

I have also added a client component to display the GitDiffReport to the user in a GitHub-like way.

### Steps for Testing

Prerequisites:
- 1 Editor or higher
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to the instructor detail page of the exercise
3. Change at least the exercise solution (Not necessary if you newly created the exercise in this PR)
4. Wait until the build is finished
5. Click on the new git-diff button
6. Check that the displayed git-diff matches the changes you did between template and solution

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| File | Branches % | Lines % |
| --- | --- | --- |
| ProgrammingExerciseGitDiffEntry | 100 | 94 |
| ProgrammingExerciseGitDiffReport | 100 | 100 |
| ProgrammingExerciseGitDiffReportService | 73 | 90 |
| ProgrammingExerciseGitDiffReportResource | 50 | 89 |
| git-diff-entry.component.ts | 76.47 | 100 |
| git-diff-modal.component.ts | 75 | 66.66 |
| git-diff-report.component.ts | 88.88 | 100 |

#### The new button:
grafik

#### The modal with the git-diff:
grafik

### CODE REVIEW

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 1]
```diff
@@ -371,7 +371,10 @@
             "diffReport": {
                 "button": "Show difference between the template and solution repositories",
                 "tooltip": "Shows and if necessary updates the git-diff between the template and solution repositories.",
-                "title": "Git-Diff"
+                "title": "Git-Diff",
+                "template": "Template",
+                "solution": "Solution",
+                "internalError": "There was an error while creating the diff. Please make sure that the template and solution repositories are properly setup."
```
Comment (ole-ve) [inline src/main/webapp/i18n/en/programmingExercise.json:377]:
```suggestion
                "internalError": "An error occured while creating the diff. Please make sure that the template and solution repositories are properly set up."

```

File: src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseGitDiffReportIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+import de.tum.in.www1.artemis.util.GitUtilService;
+import de.tum.in.www1.artemis.util.LocalRepository;
+import de.tum.in.www1.artemis.util.ModelFactory;
+
+/**
+ * Tests for the ProgrammingExerciseGitDiffReportResource and -Service
+ */
+public class ProgrammingExerciseGitDiffReportIntegrationTest extends AbstractSpringIntegrationBambooBitbucketJiraTest {
+
+    private final String fileName = "test.java";
```
Comment (ole-ve) [inline src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseGitDiffReportIntegrationTest.java:44]: Can be static, I guess. I would also separate the repository attribute declarations and other attribute declarations (fileName, templateRepo, solutionRepo, exercise).

**Review by ole-ve — DISMISSED** [review]
Tested locally, everything works as described and expected. The creation and deletion of the entities in the DB also work fine.
In addition, code changes also look good to me.
Great work, I especially like the GitHub-like-diff view 👍

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 3]
```diff
@@ -115,6 +116,10 @@
     @Column(name = "project_type", table = "programming_exercise_details")
     private ProjectType projectType;
 
+    @OneToOne(mappedBy = "programmingExercise", cascade = CascadeType.REMOVE, orphanRemoval = true, fetch = FetchType.LAZY)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:119]: add a comment that LAZY is really important here.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+            var fullEntries = new HashSet<ProgrammingExerciseFullGitDiffEntryDTO>();
+            var templateFiles = repositoryService.getFilesWithContent(templateRepo);
+            var solutionFiles = repositoryService.getFilesWithContent(solutionRepo);
+            for (ProgrammingExerciseGitDiffEntry entry : report.getEntries()) {
+                ProgrammingExerciseFullGitDiffEntryDTO fullEntry = convertToFullEntry(entry, templateFiles, solutionFiles);
+                if (!fullEntry.isEmpty()) {
+                    fullEntries.add(fullEntry);
+                }
+            }
+            fullReport.setEntries(fullEntries);
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:106]: Suggestion:
```suggestion
            var templateFiles = repositoryService.getFilesWithContent(templateRepo);
            var solutionFiles = repositoryService.getFilesWithContent(solutionRepo);
            fullReport.setEntries(report.getEntries().stream()
                    .map(entry -> convertToFullEntry(entry, templateFiles, solutionFiles))
                    .filter(fullEntry -> !fullEntry.isEmpty())
                    .collect(Collectors.toSet()));
```

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+    /**
+     * Extracts the file path from the raw git-diff for a specified diff block
+     *
+     * @param lines       All lines of the raw git-diff
+     * @param currentLine The line where the gitDiffLinePattern matched
+     * @return The file path of the current diff block
+     */
+    private String getFilePath(String[] lines, int currentLine) {
+        if (lines[currentLine - 1].startsWith("+++ ") && lines[currentLine - 2].startsWith("--- ")) {
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:350]: I suggest to add a check here, even if one thinks that it 'can't happen' :P
```suggestion
        if (currentLine > 1 && lines[currentLine - 1].startsWith("+++ ") && lines[currentLine - 2].startsWith("--- ")) {
```

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+
+    /**
+     * Extracts the previous file path from the raw git-diff for a specified diff block
+     *
+     * @param lines       All lines of the raw git-diff
+     * @param currentLine The line where the gitDiffLinePattern matched
+     * @return The previous file path of the current diff block
+     */
+    private String getPreviousFilePath(String[] lines, int currentLine) {
+        if (lines[currentLine - 1].startsWith("+++ ") && lines[currentLine - 2].startsWith("--- ")) {
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:372]: Same here
```suggestion
        if (lines > 1 && lines[currentLine - 1].startsWith("+++ ") && lines[currentLine - 2].startsWith("--- ")) {
```

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 7]
```diff
@@ -374,6 +374,13 @@
                     "pattern": "The submission limit must be and integer between 1 and 500!",
                     "required": "The submission limit must be set!"
                 }
+            },
+            "diffReport": {
+                "button": "Show difference between the template and solution repositories",
+                "tooltip": "Shows the git-diff between the template and solution repositories.™",
```
Comment (bassner) [inline src/main/webapp/i18n/en/programmingExercise.json:380]: Is hte ``™``intended? :D

File: src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts [hunk 8]
```diff
@@ -19,14 +19,25 @@ import { Task } from 'app/exercises/programming/shared/instructions-render/task/
...
 import { HttpResponse } from '@angular/common/http';
+import { ProgrammingExerciseFullGitDiffReport } from 'app/entities/hestia/programming-exercise-full-git-diff-report.model';
+import { MockNgbModalService } from '../../helpers/mocks/service/mock-ngb-modal.service';
+import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
+import { FullGitDiffReportModalComponent } from 'app/exercises/programming/hestia/git-diff-report/full-git-diff-report-modal.component';
+import { MockSyncStorage } from '../../helpers/mocks/service/mock-sync-storage.service';
+import { LocalStorageService, SessionStorageService } from 'ngx-webstorage';
+import { MockProgrammingExerciseGradingService } from '../../helpers/mocks/service/mock-programming-exercise-grading.service';
+import { ProgrammingExerciseGitDiffReport } from 'app/entities/hestia/programming-exercise-git-diff-report.model';
+import { ProgrammingExerciseGitDiffEntry } from 'app/entities/hestia/programming-exercise-git-diff-entry.model';
```
Comment (bassner) [inline src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts:30]: Unused import

File: src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts [hunk 9]
```diff
@@ -42,31 +53,66 @@ describe('ProgrammingExercise Management Detail Component', () => {
...
+                { provide: SessionStorageService, useClass: MockSyncStorage },
                 { provide: ActivatedRoute, useValue: new MockActivatedRoute() },
                 { provide: ProfileService, useValue: new MockProfileService() },
-                { provide: ProgrammingExerciseGradingService, useValue: MockProgrammingExerciseService },
+                { provide: ProgrammingExerciseGradingService, useValue: new MockProgrammingExerciseGradingService() },
+                { provide: ProgrammingExerciseService, useValue: new MockProgrammingExerciseService() },
+                { provide: NgbModal, useValue: new MockNgbModalService() },
             ],
         })
             .overrideTemplate(ProgrammingExerciseDetailComponent, '')
```
Comment (bassner) [inline src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts:86]: Can we remove this (as you're already working on it?)

File: src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts [hunk 10]
```diff
@@ -82,11 +128,14 @@ describe('ProgrammingExercise Management Detail Component', () => {
 
             // THEN
             expect(statisticsServiceStub).toHaveBeenCalled();
+            expect(gitDiffReportStub).toHaveBeenCalled();
```
Comment (bassner) [inline src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts:131]: Can we assert an absolute amount of calls instead?

File: src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts [hunk 11]
```diff
@@ -108,8 +157,11 @@ describe('ProgrammingExercise Management Detail Component', () => {
...
```
Comment (bassner) [inline src/test/javascript/spec/component/programming-exercise/programming-exercise-detail.component.spec.ts:160]: Can we assert an absolute amount of calls instead?

**Review by bassner — COMMENTED** [review]
I've had a look through the code. I marked the few things that jumped into my eye. Looks very solid overall! :)

File: src/main/webapp/i18n/en/programmingExercise.json (hunk 7 above)
Comment (Hialus) [reply src/main/webapp/i18n/en/programmingExercise.json:380]: Nope. Looks like I accidentally hit the key combination for it.

**Comment by Hialus** [comment]
Thanks for your review @bassner 🙂 
The main reason this is still a draft is that Stephan has not yet approved the database migration.
But code reviews (and local tests) are welcome.

File: src/main/java/de/tum/in/www1/artemis/web/rest/hestia/ProgrammingExerciseGitDiffReportResource.java [hunk 12]
```diff
@@ -28,34 +30,55 @@
...
 
         return ResponseEntity.ok(report);
     }
+
+    /**
+     * {@code GET exercises/:exerciseId/full-diff-report} : Create the full diff report for a programming exercise.
+     * The fill diff report contains entries with the previous and current code blocks.
+     *
+     * @param exerciseId the exerciseId of the exercise of which to create the exerciseHint
+     * @return the {@link ResponseEntity} with status {@code 201 (Created)} and with body the full diff report,
```
Comment (ole-ve) [inline src/main/java/de/tum/in/www1/artemis/web/rest/hestia/ProgrammingExerciseGitDiffReportResource.java:67]: Will this return a 201?

**Review by ole-ve — COMMENTED** [review]
Code still looks good, added some inline comments.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 13]
```diff
@@ -204,63 +297,61 @@ else if (!deactivateCodeReading) {
...
         entry.setFilePath(currentFilePath);
+        entry.setPreviousFilePath(currentPreviousFilePath);
         return entry;
     }
 
     @NotNull
-    private ProgrammingExerciseGitDiffEntry handleRemoval(List<ProgrammingExerciseGitDiffEntry> entries, String currentFilePath, ProgrammingExerciseGitDiffEntry currentEntry,
-            boolean lastLineRemoveOperation, int currentPreviousLineCount, String line) {
+    private ProgrammingExerciseGitDiffEntry handleRemoval(List<ProgrammingExerciseGitDiffEntry> entries, String currentFilePath, String currentPreviousFilePath,
+            ProgrammingExerciseGitDiffEntry currentEntry, boolean lastLineRemoveOperation, int currentPreviousLineCount, String line) {
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java]: Looks like you got an old commit, as the line parameter is not present in the current commit

**Review by ole-ve — DISMISSED** [review]
Tested locally for the following scenarios for both template and solution:
- changed the content in the same file
- deleted a file that exists in the other repository
- created a new file that does not exist in the other repository
- renamed a file
- moved a file

### What did not work:
- After the initial creation of an exercise, the diff is not generated. Only after submitting any changes to the repository, the diff is generated and therefore displayed when clicking the button.
- After deleting a file in the template or solution repository, the diff is not updated accordingly. The entries are not changed back to the initial state when re-creating this exact same deleted file. There seems to be some kind of problem in `ProgrammingExerciseGitDiffReportService::handleRemoval`. Here are the DB-entries of the wrong diff:
Bildschirmfoto 2022-03-02 um 17 27 40
(I deleted Cliente.java)

### Optional suggestions:
- In case, changes to the same file are not line-equal between the different repositories, 0 is displayed as a line number for the file the changes do not exist in. This could lead to confusion. I am not sure if this is in scope for this PR as you are using the Ace Editor, but worth mentioning I guess. In that case, not displaying the line number would be neat. 
Bildschirmfoto 2022-03-02 um 16 54 45

- In `ProgrammingExerciseGitDiffEntry` and `ProgrammingExerciseFullGitDiffEntryDTO` there is an attribute `previousFilePath` which is used for the diff on file removals. In case a file has only been moved, but not changed (except for the package-name ofc), this could also be displayed to the user. Given the fact this may be a rare use case and it's not an error in the current implementation, I suggest to tackle this (if even wanted) in a follow-up PR.

Great work, I really like the feature. 👍

**Comment by Hialus** [comment]
@ole-ve Thank you for your feedback.
I have addressed both issues in commit [11bd91d](https://github.com/ls1intum/Artemis/pull/4732/commits/11bd91d5ceda12720dee87f195284d34924bab4b).
I previously updated the git-diff every time there was a push to the repository. This, however, was not the proper way to do this, as I should have done it when the build plan finished for this push. So the git-diff is now generated after the template or solution repo build is finished. (This is required, as I rely on the test cases that are generated from this build)
The other issue when deleting files should also be fixed now.

**Review by ole-ve — DISMISSED** [review]
Again, performed all testing steps as of the previous review on [this exercise](https://artemistest.ase.in.tum.de/course-management/1211/programming-exercises/13951) with Testuser 2. I encountered the following issues:

- Adding empty lines seem to be an issue and mess up the line numbering.
Bildschirmfoto 2022-03-08 um 16 31 12

- After making changes within a file that exists in both repositories and submitting the repositories, these changes are not displayed in the git diff.

- Not sure if this is related to this PR but saving this newly created exercise after performing all the testing steps throws an Internal Server Error.

I will test this locally and take a deeper look into it.

**Comment by Hialus** [comment]
> Again, performed all testing steps as of the previous review on [this exercise](https://artemistest.ase.in.tum.de/course-management/1211/programming-exercises/13951) with Testuser 2. I encountered the following issues:
> 
>     * Adding empty lines seem to be an issue and mess up the line numbering.
>       Bildschirmfoto 2022-03-08 um 16 31 12

I don't see the issue here atm. The empty lines are of course included in the git diff as they are also changes. I will also not filter them out, as this is not the goal of this feature. This feature is only meant to create the git-diff and save it. We do not want to filter the output.

>     * After making changes within a file that exists in both repositories and submitting the repositories, these changes are not displayed in the git diff.

Not really sure what you mean by this. As seen above changes in an already existing file are shown.

>     * Not sure if this is related to this PR but saving this newly created exercise after performing all the testing steps throws an Internal Server Error.

Didn't encounter this when I tested it. If this also happens when you test it locally please send me the stack trace 🙂

**Comment by ole-ve** [comment]
I see the point of not filtering out those empty lines.

However, in the mentioned exercise, check the template repository: Client.java l.22: `// TODO: Implement everything`. This line only exists in the template and not in the solution repo but is not displayed in the diff.

**Comment by Hialus** [comment]
Thanks for clarifying. I'll look into it when I have the time.

**Comment by ole-ve** [comment]
The relevant error:
`non-transient entity has a null id: de.tum.in.www1.artemis.domain.hestia.ProgrammingExerciseGitDiffReport`
when saving the exercise.

Relevant stack trace:
ProgrammingExerciseService::updateProgrammingExercise, line 421
ProgrammingExerciseResource::updateProgrammingExercise, line 339

**Comment by Hialus** [comment]
I have now addressed both issues and also merged develop. @ole-ve

**Review by b-fein — COMMENTED** [review]
Code review part one, only server code for now.

**Review by akesfeden — DISMISSED** [review]
Tested in Testing Session on TS1, the diffs are generated correctly, also making code identical removed that file from the diff so it works as expected.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 14]
```diff
@@ -187,6 +197,17 @@ private Result processNewProgrammingExerciseResult(final ProgrammingExercisePart
         programmingSubmission.addResult(processedResult);
         programmingSubmissionRepository.save(programmingSubmission);
 
+        // Update the git-diff of the programming exercise when the push was to the solution repository and
+        // no git-diff report exists yet. This will be the case when a new exercise is created or imported
+        if ((isSolutionParticipation || isTemplateParticipation) && programmingExerciseGitDiffReportRepository.findByProgrammingExerciseId(programmingExercise.getId()) == null) {
+            try {
+                programmingExerciseGitDiffReportService.updateReport(programmingExercise);
+            }
+            catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:206]: I actually would like to keep it like this. This endpoint is not called by the Artemis client, but only by Bitbucket/Gitlab whenever a push happens.
So forwarding the message to the caller would not make that much sense.
I also want to ensure that the push is handled properly even if there was an error while generating the git-diff.

**Comment by bassner** [comment]
Is displaying an empty line with no file name, no content and no line number (just a gray block :D) intended for newly created files?
grafik

**Comment by Hialus** [comment]
Yes, that is currently intended. 
The current git-diff view is merely temporary and will be replaced in the future by a more user-friendly view.

**Review by ole-ve — APPROVED** [review]
Reviewed code, looks good since my last review. Manual testing will follow 👍

**Review by bassner — COMMENTED** [review]
I've tested the new functionality and it seems to work fine. The git diff was available on exercises generated on this PR.

A few inline comments below about the german translations, not sure how correct they are but it catched my eye

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 15]
```diff
@@ -377,6 +377,14 @@
                     "pattern": "Das Abgabelimit muss eine ganze Zahl zwischen 1 und 500 sein!",
                     "required": "Das Abgabelimit muss gesetzt sein!"
                 }
+            },
+            "diffReport": {
+                "button": "Zeige den Unterschied zwischen Template und Lösung",
+                "tooltip": "Zeigt den Git-Diff zwischen den Template und Lösung Repositories.",
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/webapp/i18n/de/programmingExercise.json:383]: Here's the thing: 
We currently use Vorlagen and Template in the German Translation.
Vorlage is used 8 times in this file and Template is used 19 times. 
Therefore I would keep Template here as it is also a German word and the closest translation from the English word. But a future PR should probably adjust this so it is consistent.

[the remark this answers is not in the document]
Comment (bassner) [reply src/main/webapp/i18n/de/programmingExercise.json]: Ah ... yeah, then we need an extra consistency PR

**Review by bassner — DISMISSED** [review]
As said above, I tested this on two test severs without issues. Code looks good to me 👍

**Review by bassner — APPROVED** [review]
Reapprove

**Review by ole-ve — APPROVED** [review]
Tested on TS1, works as described.
Code changes since my last review look good to me

**Review by canberkanar — APPROVED** [review]
The issue I mentioned is fixed. 👍🏻

**Review by akesfeden — APPROVED** [review]
Tested on TS1, git-diffs are displayed correctly, even for file name "a" and with one-line text. Code changes also look good since the last review.

**Pull request merged**